opentelemetry-api/src/opentelemetry/metrics/__init__.py

```diff
@@ -0,0 +1,245 @@
+# Copyright The OpenTelemetry Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+from re import ASCII, compile as compile_
+from typing import cast
+
+from opentelemetry.environment_variables import OTEL_PYTHON_METER_PROVIDER
+from opentelemetry.util.types import Attributes
+from abc import ABC, abstractmethod
+from logging import getLogger
+from opentelemetry.util._providers import _load_provider
+
+_logger = getLogger(__name__)
+
+
+class Measurement(ABC):
+
+    @abstractmethod
+    def __init__(self, value, **attributes: Attributes):
+        pass
+
+
+class Instrument(ABC):
+
+    _name_regex = compile_(r"[a-zA-Z][-.\w]{0,62}", ASCII)
+
+    @abstractmethod
+    def __init__(self, name, unit_of_measure="", description=""):
+
+        if self._name_regex.fullmatch(name) is None:
+            raise Exception("Invalid instrument name {}".format(name))
+
+        if len(unit_of_measure) > 63 or any(
+            ord(character) > 127 for character in unit_of_measure
+        ):
+            raise Exception("unit_of_measure must be 63 characters or shorter")
+
+
+class _Instrument(Instrument):
+
+    def __init__(self, name, unit_of_measure="", description=""):
+        super().__init__(
+            name, unit_of_measure=unit_of_measure, description=description
+        )
+
+
+class Synchronous(Instrument):
+    pass
+
+
+class Asynchronous(Instrument):
+
+    @abstractmethod
+    def __init__(self, name, callback, unit_of_measure="", description=""):
+        super().__init__(
+            name, unit_of_measure=unit_of_measure, description=description
+        )
+
+
+class Adding(Instrument):
+    pass
+
+
+class Grouping(Instrument):
+    pass
+
+
+class Monotonic(Adding):
+    pass
+
+
+class NonMonotonic(Adding):
+    pass
+
+
+class Counter(Monotonic, Synchronous):
+
+    @abstractmethod
+    def add(self, amount, **attributes):
+        if amount > 0:
```

Comment:
```suggestion
        if amount < 0:
```

Reply:
:sweat_smile: thanks for pointing that one out